scalyr_agent/builtin_monitors/linux_process_metrics.py

```diff
@@ -619,163 +732,214 @@ class ProcessMonitor(ScalyrMonitor):
+
+    def _reset_absolute_metrics(self):
+        """
+        At the beginning of each process metric calculation, the absolute metrics
+        need to be overwritten to the combined process(es) result. Only the cumulative
+        metrics need the previous value to calculate delta. We should set the
+        absolute metric to 0 in the beginning of this "epoch"
         """
-        if self.__pid is not None:
-            for gather in self.__gathers:
-                gather.close()
 
-        self.__pid = pid
-        self.__gathers = []
+        for pid, process_metrics in self.metrics_history.items():
+            for _metric, _metric_values in process_metrics.items():
+                if not _metric.is_cumulative:
+                    self.running_total_metrics[_metric] = 0
 
-        if pid is not None:
-            self.__gathers.append(StatReader(self.__pid, self.__id, self._logger))
-            self.__gathers.append(StatusReader(self.__pid, self.__id, self._logger))
-            self.__gathers.append(IoReader(self.__pid, self.__id, self._logger))
-            self.__gathers.append(FileDescriptorReader(self.__pid, self.__id, self._logger))
-        # TODO: Re-enable these if we can find a way to get them to truly report
-        # per-app statistics.
-        #        self.gathers.append(NetStatReader(self.pid, self.id, self._logger))
-        #        self.gathers.append(SockStatReader(self.pid, self.id, self._logger))
+    def _calculate_running_total(self, running_pids):
+        """
+        Calculates the running total metric values based on the current running processes
+        and the historical metric record
+        @param running_pids: list of running process ids
+        @type running_pids: list
+        """
+
+        # using the historical values, calculate the running total
+        # there are two kinds of metrics:
+        # a) cumulative metrics - only the delta of the last 2 recorded values is used (eg cpu cycles)
+        # b) absolute metrics - the last absolute value is used
+
+        running_pids_set = set(running_pids)
+
+        for pid, process_metrics in self.metrics_history.items():
+            for _metric, _metric_values in process_metrics.items():
+                if not self.running_total_metrics.get(_metric):
+                    self.running_total_metrics[_metric] = 0
+                if _metric.is_cumulative:
+                    if pid in running_pids_set:
+                        if len(_metric_values) <= 1:
+                            self.running_total_metrics[_metric] += (_metric_values[-1] if _metric_values else 0)
+                        else:
+                            self.running_total_metrics[_metric] += (_metric_values[-1] - _metric_values[-2])
+                else:
+                    # absolute metric - accumulate the last reported value
+                    self.running_total_metrics[_metric] += _metric_values[-1]
```

Comment:
I think there's a bug here if process is dead.  We only check if it is not dead for the cumulative case.